src/nsrt_learning/strips_learning/clustering_learner.py

```diff
@@ -0,0 +1,234 @@
+            for effect in pnad.op.add_effects:
+                if len(pnad.op.add_effects) > 1:
+                    # We don't want sidelining to result in a no-op.
+                    new_pnad = PartialNSRTAndDatastore(
+                        pnad.op.effect_to_side_predicate(
+                            effect, option_vars, "add"), pnad.datastore,
+                        pnad.option_spec)
+                    sprime = list(s)
+                    sprime[i] = new_pnad
+                    yield (None, tuple(sprime), 1.0)
+
+            # ...consider removing it.
+            sprime = list(s)
+            del sprime[i]
+            yield (None, tuple(sprime), 1.0)
+
+
+class ClusterAndIntersectSidelinePredictionErrorSTRIPSLearner(
+        ClusterAndIntersectSidelineSTRIPSLearner):
+    """A STRIPS learner that uses hill climbing with a prediction error score
+    function for side predicate learning."""
+
+    def _evaluate(self, initial_pnads: List[PartialNSRTAndDatastore],
+                  s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+        segments = [seg for traj in self._segmented_trajs for seg in traj]
+        strips_ops = [pnad.op for pnad in s]
+        option_specs = [pnad.option_spec for pnad in s]
+        num_true_positives, num_false_positives, _, _ = \
+            utils.count_positives_for_ops(strips_ops, option_specs, segments)
+        # Note: lower is better! We want more true positives and fewer
+        # false positives.
+        return num_false_positives + 10 * (-num_true_positives)
+
+    @property
+    def _should_satisfy_harmlessness(self) -> bool:
+        # There are no guarantees that local search to improve prediction error
+        # would satisfy harmlessness!
+        return False
+
+
+class ClusterAndIntersectSidelineHarmlessnessSTRIPSLearner(
+        ClusterAndIntersectSidelineSTRIPSLearner):
+    """A STRIPS learner that uses hill climbing with a harmlessness score
+    function for side predicate learning."""
+
+    def _evaluate(self, initial_pnads: List[PartialNSRTAndDatastore],
+                  s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+        preserves_harmlessness = self._check_harmlessness(list(s))
+        if preserves_harmlessness:
+            # If harmlessness is preserved, the score is the number of
+            # operators that we have, minus the number of side predicates.
+            # This means we prefer fewer operators and more side predicates.
+            score = 2 * len(s)
+            for pnad in s:
+                score -= len(pnad.op.side_predicates)
+        else:
+            # If harmlessness is not preserved, the score is an arbitrary
+            # constant bigger than the total number of operators at the
+            # start of the search. This is guaranteed to be worse (higher)
+            # than any score that occurs if harmlessness is preserved.
+            score = 10 * len(initial_pnads)
```

Comment:
we could make this a hyperparameter (we were using `grammar_search_{true,false}_pos_weight` before, but i think not relying on that is better...) but do we really need more?